compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -315,6 +315,7 @@ generateSrcFromLf env = noLoc mod
         Just LF.DefValue{dvalBinder=(_, ty)} <- [NM.lookup LFC.moduleImportsName . LF.moduleValues $ envMod env]
         Just quals <- [LFC.decodeModuleImports ty]
         LF.Qualified { LF.qualModule, LF.qualPackage } <- Set.toList quals
+        guard $ not $ ["DA", "Internal"] `isPrefixOf` LF.unModuleName qualModule
```

Comment:
What makes `DA.Internal` special here? I think I don’t quite understand the filter. I could see a filter for `isInternal` maybe but I’m not entirely sure that’s the right one either.

Reply:
I went with `DA.Internal` since that's the common prefix for the "missing" modules,

```
‘Pkg_9de3ae0b7b4e753f7ce2a5dcaa8d49ab4b92c4d8a8fb9589b5f8c893fbea5bd1.DA.Internal.Exception’
‘Pkg_9de3ae0b7b4e753f7ce2a5dcaa8d49ab4b92c4d8a8fb9589b5f8c893fbea5bd1.DA.Internal.Prelude’
‘Pkg_9de3ae0b7b4e753f7ce2a5dcaa8d49ab4b92c4d8a8fb9589b5f8c893fbea5bd1.DA.Internal.Template.Functions’
```

Reply:
Right what I was trying to understand is what makes these modules special, i.e., why do we pull those in when we shouldn’t and not any other modules?

Reply:
but yes, maybe `isInternal` would make more sense. The reason we end up with these three is that they define at least one orphan instance, but the corresponding `CurrentSdk` instance should be available anyway since these are imported implicitly, right?

Reply:
I think I still don’t fully understand it, let me try to walk through it and you point out where my understanding is off (probably just need more coffee):

1. The original module ends up importing these modules even if there are no templates. I think there is no direct import but it’s imported transitively via prelude.
2. These modules define orphan instances. That means we record them in LF so we can reconstruct those imports in data-dependencies.
3. So now we get to reconstruction in data-dependencies. We see that there we had orphan instances so we try to generate an import for it. That fails because the module is not there.

Is that right so far?

So first a question: Is `9de3ae0…` the package id of the old daml-stdlib or the new daml-stdlib? I assume the old but not entirely sure.

And now to the parts that I don’t understand:
1. Why are those modules not in scope? Shouldn’t we bring them in scope if we have a reference to them via orphan instances? It sounds like our dependency tracking might be slightly off here and if it is I’m not quite sure it really only affects these modules.
2. Why does it work if we actually define a template? Clearly adding more code does not make us reference those modules any less. So it seems like in that case, we actually track the reference somehow and produce it correctly?

Reply:
Thanks for sharing your thought process. The first three match my thinking exactly. 

> So first a question: Is 9de3ae0… the package id of the old daml-stdlib or the new daml-stdlib? I assume the old but not entirely sure.

It must be the old one since it's part of the `$$imports` value for module `A`

> 2. Why does it work if we actually define a template? Clearly adding more code does not make us reference those modules any less. So it seems like in that case, we actually track the reference somehow and produce it correctly?

I found that using *any* type or value from `daml-stdlib` in `A` makes `B` work, e.g. `identity` or `undefined`. AFAICT anything where `daml damlc inspect` shows a type or value from `daml-stdlib` makes the problem go away, it doesn't have to be a template. 

> 1. Why are those modules not in scope? Shouldn’t we bring them in scope if we have a reference to them via orphan instances? It sounds like our dependency tracking might be slightly off here and if it is I’m not quite sure it really only affects these modules.

indeed, it seems that we are not tracking references via unused instances. And your intuition is right, I was able to get a similar error with an orphan instance for a type and a class outside `daml-stdlib`, 

```
File:
  .daml/package-database/1.14/a-0.0.1-dc295f9f9c095341292713f3312c4340c9a6644f9bab916e97f247a5242439f3/A.daml
Hidden:   no
Range:    21:8-21:79
Source:   not found
Severity: DsError
Message:
  Could not find module ‘Pkg_bcbdca2d32dc9f7b2d9ed0c88ee04509a72f0aa4a752473e85ffa79eadbf2abe.ZI’
  It is not a module in the current program, or in any known package.
Failed to compile interface for data-dependency: a-0.0.1
```

see [data-deps-orphans.tar.gz](https://github.com/digital-asset/daml/files/8131797/data-deps-orphans.tar.gz), so this definitely won't fix that. Interestingly, `daml damlc inspect-dar` on the `a` `dar` shows that the `z` package _is_ there, which should make things easier. I'll try to come up with a different approach.

Reply:
I think the issue is that `packageRefs`, as used in `DA.Cli.Damlc.Packaging.buildLfPackageGraph` (which is where data-dependencies gets called), doesn't take into account the `$$import`, it only tracks references in the LF AST.

Not sure what the best solution there is. Maybe instead of using `packageRefs` there, we should use something that also looks at `$$import` values?

Reply:
Maybe this is a good reason for us to actually add imports into the LF AST, even if they don't do anything.

Reply:
@sofiafaro-da thanks for pointing me to `buildLfPackageGraph`. I do think this is getting close to the root of the issue. I'm going to see if adding the `$$import`s to `packageRefs` fixes both errors (the one from #13003 and the one from my previous comment), and later we can discuss about adding imports to the LF AST